libs/wire-api-federation/proto/router.proto

```diff
@@ -0,0 +1,74 @@
+syntax = "proto3";
+package wire.federator;
+
+// These definitions are part of the federation protocol.
+//
+// A component (e.g. 'brig') will send some data of type 'RemoteCall' to
+// the local 'federator' service. That federator will extract the domain from
+// the 'RemoteCall' to discover where to send some data of type 'LocalCall' to.
+// On the other end, the 'LocalCall' will be converted to a 'Response'.
+
+enum Component {
+  Brig = 0;
+}
+
+message HTTPResponse {
+    uint32 responseStatus = 1;
+    bytes responseBody = 2;
+}
+
+message Response {
+  oneof response {
+    HTTPResponse httpResponse = 1;
+    string err = 2;
+  }
+}
+
+// the envelope thing which is sent from brig to local federator
+message RemoteCall {
+  string domain = 1;
+  LocalCall localCall = 2;
+}
+
+// the federator will take out LocalCall from the RemoteCall Envelope, and
+// forward the LocalCall object to a federator on another backend (with the wish
+// that this remote federator forwards this to its own local brig)
+message LocalCall {
+  Component component = 1;
+  Method method = 2;
+  bytes path = 3;
+  repeated QueryParam query = 4;
+  bytes body = 5;
```

Comment:
So GRPC is use to encode arbitrary HTTP requests?  Will it be feasible to elegantly encode GRPC requests to brig once that is a thing, rather than rest/json?

I was expecting to see a higher-level protocol encoded here, but I'm beginning to feel this may have been a better idea.  I should probably read the documentation and see if there is a discussion about this giving the reasons for the decision.  Feel free to drop me a link, though.  :)

Reply:
> So GRPC is use to encode arbitrary HTTP requests? Will it be feasible to elegantly encode GRPC requests to brig once that is a thing, rather than rest/json?

Yes, it should be fairly simple as in GRPC you just have to say name of the service, name of the procedure and pass the data in. Supporting streaming services might be more interesting, but I can't think of any use case for streaming services.

> I was expecting to see a higher-level protocol encoded here, but I'm beginning to feel this may have been a better idea. I should probably read the documentation and see if there is a discussion about this giving the reasons for the decision. Feel free to drop me a link, though. :)

Initially we were going to encode higher level protocol here, if by that you mean stuff like "getUserByHandle(handle)". But more I thought about it more it didn't make sense to have to repeat all of this everywhere (serving brig, calling from brig, serving from here and calling from here). Plus that would add testing burden, as in for each thing we add we won't be sure it works until it is tested across two backends. This seemed like we were tangling routing concerns (find a remote and talk to it) with application logic concerns (getUserByHandle), which I think was opposite of why we wanted a federator. Trying to separate these things out until writing federated code was as simple as possible got us here (It can still be simpler, but its a start). Unfortunately, we didn't write anything about this, we plan to do this soon. We're hoping that it will go in the `docs/reference` directory.

Reply:
I added some comments to explain in https://github.com/wireapp/wire-server/pull/1319/commits/c2667aef77b3d5ad6868f6a6fd80bbe9f47fa12b - Akshay, feel free to add something there, or to docs/reference/federation.